### PR TITLE
Blog-App: Validation feature

### PR DESCRIPTION
**Pull request by tajulafreen**
**In this PR I made the following changes:**

- Add the following validations:
  - For the User model:
  - Name must not be blank.
- PostsCounter must be an integer greater than or equal to zero.
- For the Post model:
  The title must not be blank.
  The title must not exceed 250 characters.
- CommentsCounter must be an integer greater than or equal to zero.
- LikesCounter must be an integer greater than or equal to zero.
- Add unit specs for all of your models' methods and validations.

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -3,6 +3,9 @@ class Post < ApplicationRecord
   has_many :comments
   has_many :likes
 
+  validates :title, presence: true, length: { maximum: 250 }
+  validates :comments_counter, :likes_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+
```
Comment (mckent05) [inline app/models/post.rb:8]: - You have done well so far, please validate that in your post model, the LikesCounter must be an integer greater than or equal to zero. Please fix this 😄

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    expect(post).to_not be_valid
+  end
+
+  it 'increments the user posts_counter after creation' do
+    user = User.create(name: 'Charlie', posts_counter: 0)
+    user.posts.create(title: 'New Post', comments_counter: 0, likes_counter: 0)
+    user.reload
+    expect(user.posts_counter).to eq(1)
+  end
+end
```
Comment (mckent05) [inline spec/models/post_spec.rb]: - You have done well so far, but according to the project requirements, Any custom methods in the models should have a corresponding unit test. Please test for the `recent_comments` method in your post model 👍 👍

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+  it 'is invalid without a name' do
+    user = User.new(posts_counter: 5)
+    expect(user).not_to be_valid
+  end
+
+  it 'is invalid with a negative posts_counter' do
+    user = User.new(name: 'John', posts_counter: -1)
+    expect(user).not_to be_valid
+  end
```
Comment (mckent05) [inline spec/models/user_spec.rb:17]: - You have done well so far, but according to the project requirements, Any custom methods in the models should have a corresponding unit test. Please test for the `recent_posts` method in your user model 👍 👍. Please fix this 😄

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @tajulafreen, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍 
- Good job adding validation and testing your model 👏 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/post_spec.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    post.comments.create(author: user, body: 'Old comment')
+    newer_comment1 = post.comments.create(author: user, body: 'Newer comment 1')
+    newer_comment2 = post.comments.create(author: user, body: 'Newer comment 2')
+    newer_comment3 = post.comments.create(author: user, body: 'Newer comment 3')
+    newer_comment4 = post.comments.create(author: user, body: 'Newer comment 4')
+    newer_comment5 = post.comments.create(author: user, body: 'Newer comment 5')
+
+    recent_comments = post.recent_comments
+
+    expect(recent_comments).to eq([newer_comment5, newer_comment4, newer_comment3, newer_comment2, newer_comment1])
```
Comment (Munsa1) [inline spec/models/post_spec.rb]: - [ ] Good job implementing the recent comments and recent posts tests, however your tests are not passing, kindly fix this
image

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @tajulafreen, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
Added Tests As Requested :heavy_check_mark: 


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by t-yanick — APPROVED** [review]
# STATUS: MILESTONE APPROVED :1st_place_medal: 

Hi @tajulafreen , 

### Great job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Requested change implemented 👍 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**